### PR TITLE
Change upload API - client sends files and hashes (BL-12568)

### PR DESCRIPTION
**Pull request by andrew-polk**
This change is [Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)

### CODE REVIEW

**Comment by andrew-polk** [comment]
Draft because it must be coordinated with the BloomDesktop side.

**Review by andrew-polk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: 0 of 2 files reviewed, 1 unresolved discussion

___
*[`book/uploadStart.ts` line 190 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoNF5iCDTml2GyD6GuH:-NoNF5iCDTml2GyD6GuI:b-9eg8oy) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L190)):*
> ```XML
>       try {
>         // existingBookPath is in the form bookId/timestamp/title.
>         // prefix only has bookId/timestamp. The client must provide the new title; it could have changed.
> ```

I'm unsure about the best way to handle this title prefix.
I would really prefer the API not have to include it, but I haven't thought of a better option.
In theory, we could just add it to the beginning of each file name, but
1. That is extra bits transmitted
2. We would have to know to strip that off when checking the files against existing S3 anyway. So there is still special knowledge of it either way.

I've chosen to call it a prefix for now because I didn't want to specify it is a title in the API. We think we would like to get away from having the title in the path some day.
But the name is kinda confusing since we use `prefix` for the bookId/timestamp portion.

**Review by hatton — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @andrew-polk)

___
*[`book/uploadStart.ts` line 190 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoNF5iCDTml2GyD6GuH:-NoOHARZCRBbRY_4h-7O:bho6i2l) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L190)):*
I'd worry less about bits than about round-trips and do whatever is cleaner. But I don't find the prefix thing hard to follow, just need to liberally comment that it's actually the title. 

> But the name is kinda confusing since we use `prefix` for the bookId/timestamp portion.

This part seems concerning but I don't know where that is. So if you have two "prefixes", and can't give them better names, then it's better to just let one be "title". Sorry, it's hard to be helpful from afar.

**Review by JohnThomson — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @andrew-polk and @hatton)

___
*[`book/uploadStart.ts` line 124 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoOkkBc615sar2OL474:-NoOkkBc615sar2OL475:b-yexm4h) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L124)):*
> ```XML
>   let filesToUpload: string[] = [];
>   if (!isNewBook) {
>     // we are modifying an existing book. Check that we have permission, then copy old book to new folder for efficient syncing
> ```

Comment needs tweaking, I think. The following code does not copy files, just deletes any incomplete upload.
What would delete an incomplete upload of a new book? Or would it no longer be considered new if it had even an incomplete upload? How would that relate to the Bloom UI asking whether to replace it, if it is partly uploaded but there is no complete prior upload?
___
*[`book/uploadStart.ts` line 156 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoOl_mcAgDoSvjhQIgh:-NoOl_mcAgDoSvjhQIgi:b-94huz4) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L156)):*
> ```XML
>       const filesToDelete = allFilesForThisBook.filter(
>         (file) => !currentlyUsedFilesForThisBook.includes(file)
>       );
> ```

Feels like somewhere about here it would be really easy to just NOT delete any file we find to have a desired name and checksum, and instead remove it from the list of things Bloom must upload or that must be copied from the previous version. Restartable upload in about a dozen lines of code!
I suppose you'd also have to pass the list of keepers to processFileHashes so it could avoid putting them in either result. Or maybe it would be cleaner to let processFileHashes consider the new directory also, and return three lists: copy, upload, and delete, with a first preference for finding the right file with the right hashcode already in the destination, and leaving it out of all three results.
___
*[`book/uploadStart.ts` line 190 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoNF5iCDTml2GyD6GuH:-NoOfTHT7Mlg-lYeBybc:b-ormab2) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L190)):*
<details><summary><i>Previously, hatton (John Hatton) wrote…</i></summary><blockquote>

I'd worry less about bits than about round-trips and do whatever is cleaner. But I don't find the prefix thing hard to follow, just need to liberally comment that it's actually the title. 

> But the name is kinda confusing since we use `prefix` for the bookId/timestamp portion.

This part seems concerning but I don't know where that is. So if you have two "prefixes", and can't give them better names, then it's better to just let one be "title". Sorry, it's hard to be helpful from afar.
</blockquote></details>

So we're still keeping BloomLibraryBooks/<uploader email>/<bookId>, adding <timestamp> to distinguish different uploads of the same book, and then keeping <title> so at least you can see at a glance what book it is. But having the title in there is something of a pain when it changes, though there will always only be one book (and perhaps the beginnings of a new upload) under BloomLibraryBooks/<uploader email>/<bookId>. (Is all that right?)
From the upload API point of view, therefore, it doesn't matter whether the extra bit of  prefix is the book title. It could be the book guid again. I think it would even work fine if it was always an empty string (though I very vaguely recall we might use the last name element as a progress dialog title when downloading??)
Since it's main purpose is to make the book recognizable to human browsers of S3 (and perhaps in the download dialog) , what if we called the last element something like humanFriendlyName or readableBookId or even just folderName? Or some recombination of those ideas. It happens to be USED as part of an S3 prefix, but names should be more about purpose, especially since it's only part of the S3 prefix. 
Note that if we think of it just as a human-friendly name to help identify the book, we could one day do other things with it besides use it as part of the prefix. For example, we could store it in an S3 file in the book folder called "humanFriendlyName.txt", or ignore it altogether, or save it in the parse database.
___
*[`common/s3.ts` line 271 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoOo-xu7qwQxI7KYqUa:-NoOo-xu7qwQxI7KYqUb:br0d7cx) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/common/s3.ts#L271)):*
> ```XML
>     filesNewOrModified = clientFiles.map((file) => file.path);
>     return [filesNewOrModified, filesNotChanged];
>   }
> ```

I have a vague recollection that you were worried about some limit on command line length (doesn't seem to apply after all) or max number of results from a query (might be a problem). Did all that turn out not to be an issue?

**Review by andrew-polk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @hatton and @JohnThomson)

___
*[`book/uploadStart.ts` line 124 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoOkkBc615sar2OL474:-NoYGTUwEH7FNe7OzYaS:b-zcjvte) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L124)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Comment needs tweaking, I think. The following code does not copy files, just deletes any incomplete upload.
What would delete an incomplete upload of a new book? Or would it no longer be considered new if it had even an incomplete upload? How would that relate to the Bloom UI asking whether to replace it, if it is partly uploaded but there is no complete prior upload?
</blockquote></details>

Thanks
___
*[`book/uploadStart.ts` line 156 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoOl_mcAgDoSvjhQIgh:-NoYOsHqD1td0uSiUOFR:b-q7m2bw) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L156)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Feels like somewhere about here it would be really easy to just NOT delete any file we find to have a desired name and checksum, and instead remove it from the list of things Bloom must upload or that must be copied from the previous version. Restartable upload in about a dozen lines of code!
I suppose you'd also have to pass the list of keepers to processFileHashes so it could avoid putting them in either result. Or maybe it would be cleaner to let processFileHashes consider the new directory also, and return three lists: copy, upload, and delete, with a first preference for finding the right file with the right hashcode already in the destination, and leaving it out of all three results.
</blockquote></details>

I agree it should be pretty doable.
___
*[`book/uploadStart.ts` line 190 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoNF5iCDTml2GyD6GuH:-NoYFUXPB-Ylpx6t45ax:bdm1kom) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L190)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

So we're still keeping BloomLibraryBooks/<uploader email>/<bookId>, adding <timestamp> to distinguish different uploads of the same book, and then keeping <title> so at least you can see at a glance what book it is. But having the title in there is something of a pain when it changes, though there will always only be one book (and perhaps the beginnings of a new upload) under BloomLibraryBooks/<uploader email>/<bookId>. (Is all that right?)
From the upload API point of view, therefore, it doesn't matter whether the extra bit of  prefix is the book title. It could be the book guid again. I think it would even work fine if it was always an empty string (though I very vaguely recall we might use the last name element as a progress dialog title when downloading??)
Since it's main purpose is to make the book recognizable to human browsers of S3 (and perhaps in the download dialog) , what if we called the last element something like humanFriendlyName or readableBookId or even just folderName? Or some recombination of those ideas. It happens to be USED as part of an S3 prefix, but names should be more about purpose, especially since it's only part of the S3 prefix. 
Note that if we think of it just as a human-friendly name to help identify the book, we could one day do other things with it besides use it as part of the prefix. For example, we could store it in an S3 file in the book folder called "humanFriendlyName.txt", or ignore it altogether, or save it in the parse database.
</blockquote></details>

I don't know about the reason behind including title in the first place.
We have to maintain it (or at least the same number of levels) now for older Blooms to continue downloading. And yes, it is used by the download dialog to display the book title.

I've reversed my original inclination against using title and decided to just call it `title` because that's what it is. The server can decide what, if anything, to do with it. Naming it something else in the API actually puts an assumption on *how* it will be used which is what I was trying to get away from.

(but I'm open to further discussion)
___
*[`common/s3.ts` line 271 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoOo-xu7qwQxI7KYqUa:-NoYFDPU0zQbawkl0G4V:b-ujkdqg) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/common/s3.ts#L271)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I have a vague recollection that you were worried about some limit on command line length (doesn't seem to apply after all) or max number of results from a query (might be a problem). Did all that turn out not to be an issue?
</blockquote></details>

We were looking at doing a copy by prefix and there were limitations on exclusions (so we didn't copy the files which were being deleted). Since we can't do copy by prefix, it is now irrelevant.

**Comment by andrew-polk** [comment]
*[`book/uploadStart.ts` line 190 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoNF5iCDTml2GyD6GuH:-NoZ1wBk-SMOj0aJkFDV:buft0m7) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L190)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I don't know about the reason behind including title in the first place.
We have to maintain it (or at least the same number of levels) now for older Blooms to continue downloading. And yes, it is used by the download dialog to display the book title.

I've reversed my original inclination against using title and decided to just call it `title` because that's what it is. The server can decide what, if anything, to do with it. Naming it something else in the API actually puts an assumption on *how* it will be used which is what I was trying to get away from.

(but I'm open to further discussion)
</blockquote></details>

Actually, looks like I was wrong.
We don't use the title from the S3 prefix for displaying the title in the download dialog. A long time ago, we started putting it as a query param in the bookOrderUrl.

But older Blooms do use the title out of the S3 prefix to generate the name of the folder where it downloads the files under Books for BloomLibrary.org.

**Comment by andrew-polk** [comment]
*[`book/uploadStart.ts` line 190 at r1](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoNF5iCDTml2GyD6GuH:-NoZ50Em1xP2FVISTT0V:b6tjxaw) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/229b3a9cf8001d24d33dcde6c8f7c0b0f4d854f7/book/uploadStart.ts#L190)):*
> But older Blooms do use the title out of the S3 prefix to generate the name of the folder where it downloads the files under Books for BloomLibrary.org.

And current Blooms for that matter. That's what I'm looking into simplifying at the moment.

**Review by JohnThomson — APPROVED** [review]
I think this looks good. One minor suggestion, but it's only an optimization.

> Reviewed 1 of 2 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-:-NoqnvvuD9AVAP2SPdYS:b-l5oxjs)* status: all files reviewed, 1 unresolved discussion (waiting on @andrew-polk)

___
*[`book/uploadStart.ts` line 169 at r2](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoqnYPT6uBfdscqvm7E:-NoqnYPT6uBfdscqvm7F:b6v598b) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/db4ca709d9aba87ce33737e421b68fd973408eed/book/uploadStart.ts#L169)):*
> ```XML
>       const filesToDelete = allFilesForThisBook.filter(
>         (file) => !currentlyUsedFilesForThisBook.includes(file)
>       );
> ```

This may be premature or insignificant optimization, but can't we just look at the files returned by the first query and eliminate any that start with existingBookPath?

**Review by andrew-polk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: all files reviewed, 1 unresolved discussion (waiting on @JohnThomson)

___
*[`book/uploadStart.ts` line 169 at r2](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NoqnYPT6uBfdscqvm7E:-NorbOfU7l2cN9hDnn7_:bhvjwi4) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/db4ca709d9aba87ce33737e421b68fd973408eed/book/uploadStart.ts#L169)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This may be premature or insignificant optimization, but can't we just look at the files returned by the first query and eliminate any that start with existingBookPath?
</blockquote></details>

Thanks. I think I changed it to be much cleaner.
Also, there was a bug waiting because the prior code didn't handle more than 1000 files.

**Review by JohnThomson — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: all files reviewed, 2 unresolved discussions (waiting on @andrew-polk)

___
*[`book/uploadStart.ts` line 160 at r3](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NosII_X-luUAmd_Di9w:-NosII_X-luUAmd_Di9x:b-qoqhvv) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/c209f72b570ca4d946e08ff1f523442af08d5ad7/book/uploadStart.ts#L160)):*
> ```XML
>     if (existingBookInfo.uploadPendingTimestamp) {
>       try {
>         await deleteBook(bookObjectId, env, existingBookPath);
> ```

This looks to me as though it will delete not the incomplete upload, but any previous complete upload...which I don't think we want to do until we have a new, successful one, right? Don't we need to come up with a prefix for deleting that involves the bit without a timestamp plus the uploadPendingTimestamp? Which I would expect to be different from the timestamp that is part of existingBookPath and refers to a previous upload?

**Review by andrew-polk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @JohnThomson)

___
*[`book/uploadStart.ts` line 160 at r3](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-NosII_X-luUAmd_Di9w:-NosQJ1X-gPQ1qmmZKvi:b-58nvw0) ([raw file](https://github.com/BloomBooks/bloom-azure-functions/blob/c209f72b570ca4d946e08ff1f523442af08d5ad7/book/uploadStart.ts#L160)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This looks to me as though it will delete not the incomplete upload, but any previous complete upload...which I don't think we want to do until we have a new, successful one, right? Don't we need to come up with a prefix for deleting that involves the bit without a timestamp plus the uploadPendingTimestamp? Which I would expect to be different from the timestamp that is part of existingBookPath and refers to a previous upload?
</blockquote></details>

I've attempted to clarify by changing names and comments.
See what you think.

**Review by JohnThomson — APPROVED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72)* status: all files reviewed, 1 unresolved discussion (waiting on @andrew-polk)

**Review by JohnThomson — APPROVED** [review]
:lgtm:

> Reviewed 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/bloom-azure-functions/72#-:-Np0j2Fh81IOMkVU6f6i:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @andrew-polk)
